### PR TITLE
auth: add key support in merge func

### PR DESCRIPTION
**Pull request by xiang90**
/cc @heyitsanthony @mitake 

To combine the range checking and key checking code path, we need to treat key, range the same in the merge function.

### CODE REVIEW

File: auth/range_perm_cache.go [hunk 1]
```diff
@@ -22,9 +22,20 @@ import (
...
-	return 0 <= strings.Compare(a.begin, b.begin) && strings.Compare(a.end, b.end) <= 0
+	switch {
+	case len(a.end) == 0 && len(b.end) == 0:
+		// a, b are both keys
+		return strings.Compare(a.begin, b.begin) == 0
+	case len(b.end) == 0:
+		// b is a key, a is a range
+		return false
+	case len(a.end) == 0:
+		return 0 <= strings.Compare(a.begin, b.begin) && strings.Compare(a.begin, b.end) <= 0
```
Comment (heyitsanthony) [inline auth/range_perm_cache.go]: are `rangePerm`s  `[begin, end)`?

Comment (xiang90) [inline auth/range_perm_cache.go]: right now, it is [begin, end]. I plan to clean this up later.

**Comment by heyitsanthony** [comment]
lgtm

**Comment by mitake** [comment]
@xiang90 lgtm, thanks a lot

**Pull request merged**